Added support for time in dateinput component

shivankacker opened the pull request:
## Proposed Changes

- Fixes #6858 
- Adds support for time in the `DateInputV2` component.
- Replaces all occurrences of datetime-local inputs

![image](https://github.com/user-attachments/assets/c8b24e95-109d-4a9a-ba78-3c72f7b50912)

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

shivankacker committed:
Added support for time in dateinput

shivankacker committed:
made responsive

shivankacker committed:
fixed responsiveness and updated tests

shivankacker committed:
Merge branch 'develop' of https://github.com/ohcnetwork/care_fe into support-time-in-dateinput

shivankacker committed:
Fixed scroll behaviour

shivankacker committed:
minor test fix

nihal467 commented:
@shivankacker 

![image](https://github.com/user-attachments/assets/5c0fffe9-e18a-4d99-aab4-ddb69e28c4f2)

- [ ] When you open the date and time component in the patient consultation page, the entire form is scrolled to the right side, making it hard for the user to see what is the field heading since they are left-aligned
- [ ] Once a user clicks on a specific time, then the component should auto-close

![image](https://github.com/user-attachments/assets/e5fb6deb-064c-4f63-b059-401a763977b8)

- [ ] the placeholder text { hh: mm a }, what is meant by a in this text?

![image](https://github.com/user-attachments/assets/9d54879c-1da9-411c-b723-8ce1d5093d6d)

- [ ] We had an existing behavior where typing "12092024" would automatically format the input as 12/09/2024. This behavior was removed in your PR. Please revert the change and retain the auto-formatting in the component.
- [ ] When the user types "12092024" and moves out of focus of the element or switches tabs, the entered data resets to the current date and time in the discharge pop-up.

![image](https://github.com/user-attachments/assets/990589f0-5a95-419a-a403-c5c59965bf4c)

- [ ] On mobile view, the component is clipping in, making it properly responsive across devices

![image](https://github.com/user-attachments/assets/888ef44e-0c30-4f32-8d08-6b4dec123045)

- [ ] Since the auto-formatting is not working, there is no limit in user input

![image](https://github.com/user-attachments/assets/e2b473dd-24e8-4451-bdb3-1e00c8eee10b)

- [ ] even if user manually, type all the details by manually formatting, the component still reset to current date and time

CC: @gigincg

nihal467 commented:
@shivankacker Bro, create a checklist of all the forms where you're changing the component, and add it to the description. Review the functionality and responsiveness of each form one by one before the next QA review.

shivankacker commented:
@nihal467 

1. Will see to the scroll issue. Wasn't happening with my screen size
2. I did not keep the auto closing thing because : It will close if a user will set the time before setting the date or vice-versa which adds makes it a UX issue.
3. "a" means AM/PM. If it is not understandable I will change it to "am/pm"
> We had an existing behavior where typing "12092024" would automatically format the input as 12/09/2024
4. The existing behaviour made it easy to enter the date, but totally hard to update it. The reason it was removed was so that the user does not face this while editing the date:

https://github.com/user-attachments/assets/7313c037-a439-4662-9cf0-5c038ecc50a4

Because this behaviour was changed, the tests were also changed.

5. The input will auto reset if the entered time is not of the desired format.
6. On mobile devices, the component can be scrolled left-right. This is because the date input would become small if i try to change sizes/ do anything else.
7. You did not add space, which made it an invalid format.

shivankacker committed:
Updated placeholder, changed layout of time picker on mobile, and made some fixes for very edge cases

shivankacker commented:
@nihal467 
- because you did not like the left-right scroll, I changed it to go the next row. Let me know if it doesn't look good.
- changed "a" to "am/pm"
- fixed the edge case of the overflow. Also verified all the inputs again. Please let me know if I missed out on something.

cc. @gigincg

shivankacker committed:
Made fool proof

shivankacker commented:
Also the input will show a red border now if the input format is wrong. Just an add on.

shivankacker committed:
Merge branch 'develop' of https://github.com/ohcnetwork/care_fe into support-time-in-dateinput

shivankacker committed:
fix merge conflicts

shivankacker committed:
stashing

shivankacker committed:
fix merge conflicts

bodhish commented:
@shivankacker can you update the status here?

shivankacker commented:
@bodhish still working on this, waiting to complete other priority issues. Will be covered by this sprint.

shivankacker committed:
fix merge conflicts

shivankacker committed:
updated UI to not have the user add slashes

shivankacker committed:
updated tests

shivankacker committed:
Merge branch 'develop' of https://github.com/ohcnetwork/care_fe into support-time-in-dateinput

shivankacker committed:
fixed tests

shivankacker committed:
Merge branch 'develop' of https://github.com/ohcnetwork/care_fe into support-time-in-dateinput

shivankacker committed:
fix responsiveness

nihal467 commented:
![image](https://github.com/user-attachments/assets/6c170d02-ed2b-4f17-8d24-f505d054953d)

- [ ] the component is in 12 hours clock by default, even though the system clock is in 24 hours

shivankacker committed:
fix merge conflicts

shivankacker committed:
fixed bugs

shivankacker commented:
@nihal467
- fixed the size bug
- Updated placeholder

The text is in 24 hour format so that the user can just input numbers and does not have to input AM/PM through the keyboard. Also reduces the code complexity. We can take it up later if any issue arises from the field.

nihal467 commented on cypress/pageobject/Asset/AssetCreation.ts:
```diff
@@ -62,7 +62,10 @@ export class AssetPage {
+    cy.get('[data-test-id="date-input"]:visible [data-time-input="0"]')
+      .click(0, 0)
+      .type(lastServicedOn);
+    cy.get("body").click(0, 0);
```
why are you using the hardcoded position, rather try using cy.clickAndTypeDate() command

shivankacker replied:
Hard coded position is only used to click outside of the popup

nihal467 commented on cypress/pageobject/Asset/AssetCreation.ts:
```diff
@@ -120,7 +123,13 @@ export class AssetPage {
+    cy.get('[data-test-id="date-input"]:visible [data-time-input="0"]')
+      .click()
+      .type(lastServicedOn);
+    cy.get("body").click(0, 0);
```
here also, try the command function mention above

nihal467 commented on cypress/pageobject/Asset/AssetCreation.ts:
```diff
@@ -268,7 +277,13 @@ export class AssetPage {
+    cy.get('[data-test-id="date-input"]:visible [data-time-input="0"]')
+      .click()
+      .type(text);
+    cy.get("body").click(0, 0);
```
here also, try the command function mention above

nihal467 commented on cypress/pageobject/Facility/FacilityCreation.ts:
```diff
@@ -246,7 +246,10 @@ class FacilityPage {
+    cy.get('[data-test-id="date-input"]:visible [data-time-input="0"]')
+      .click()
+      .type(date);
+    cy.get("body").click(0, 0);
```
here also, try the command function mention above

nihal467 commented on cypress/pageobject/Patient/PatientConsultation.ts:
```diff
@@ -61,7 +61,14 @@ export class PatientConsultationPage {
+    cy.get('[data-test-id="date-input"]:visible [data-time-input="0"]')
+      .click()
+      .type(date);
+    cy.get("body").click(0, 0);
```
here also, try the command function mention above

nihal467 reviewed (dismissed):
- [x] dont use hardcoded position based click in cypress test
- [x] try to use cy.clickAndTypeDate() command or something reusable

bodhish commented:
@shivankacker I want this merged ASAP

shivankacker committed:
fixed merge conflicts and updated tests

shivankacker committed:
little mistake

shivankacker commented:
@nihal467 switched to helper function wherever applicable. The only place hardcoded position is being used is to make sure the user has clicked outside the popup.

shivankacker committed:
autoclose on first date input

shivankacker committed:
fix merge conflicts

shivankacker committed:
fix merge conflicts

bodhish commented on package-lock.json:
```diff
@@ -11974,6 +11974,23 @@
         "loose-envify": "cli.js"
       }
     },
+    "node_modules/lottie-react": {
```
why do we have change in the lock?

shivankacker replied:
might have slipped in with the latest merge conflicts, sorry about that

bodhish commented on src/components/Common/DateInputV2.tsx:
```diff
@@ -1,17 +1,20 @@
   | "RIGHT"
   | "CENTER"
+  | "LEFT-CENTER"
+  | "RIGHT-CENTER"
```
where are you using it?

bodhish commented on src/components/Common/DateInputV2.tsx:
```diff
@@ -25,12 +28,14 @@ interface Props {
   placeholder?: string;
   isOpen?: boolean;
   setIsOpen?: (isOpen: boolean) => void;
+  time?: boolean;
```
what does `time: true` mean? not sure if the variable naming is right.

bodhish commented on src/components/Common/DateInputV2.tsx:
```diff
@@ -246,238 +297,357 @@ const DateInputV2: React.FC<Props> = ({
-                                key={day}
-                                id={`day-${i}`}
-                                className="aspect-square w-[14.26%]"
+                                onClick={showYearPicker}
```
why not set the state here?

bodhish reviewed (changes requested):
Only covered 30% @shivankacker self review your code once.

shivankacker committed:
cleanup

shivankacker commented:
@bodhish cleaned up as much as possible

bodhish reviewed (approved):
Lets merge this and see what breaks!

bodhish commented:
@shivankacker do fix the conflicts and merge it asap.

shivankacker committed:
fixed merge conflicts

shivankacker commented:
@bodhish can you merge before another merge conflict happens 🥲

shivankacker commented:
![image](https://github.com/user-attachments/assets/c7cd8d08-380f-449c-b2e3-2677f7617f48)
